### ISSUE TITLE
Data type returned by HTTP is not supported

### ISSUE THREAD

**Issue by manks0127** [yiisoft/yii2-authclient#148]
In the 2.1 version, I tried to get the QQ platform OpenID through the baseOAuth method of sendRequest, but yii2-httpclient's getParser method throws an exception like
Unable to format for parsing. Raw content detect response:
Http-Code: 200
TWS Server:
Date: Tue, 03 Jan 2017 15:08:16 GMT
Text/html Content-Type:
Content-Length: 83
Close Connection:
No-cache Cache-Control:
Callback ({"client_id": "101318901", "openid" "F2FDA4A1E262F956E37F47B767958A65"});

I'm trying to analyze what causes it.
Finally, it is found that the yii2-httpclient detectFormatByContent does not support the return of the content type
In the 2.0 version of the BaseOAuth sendRequest method, I can override the processResponse method to deal with such data content, but in the 2.1, how to solve this problem?

In China, QQ is the largest chat software, very many people in use, I think the above problems should be paid attention to! The following QQ related documents: http://wiki.open.qq.com/wiki/website/%E8%8E%B7%E5%8F%96%E7%94%A8%E6%88%B7OpenID_OAuth2.0

**Comment by manks0127** [comment]
@yii-bot @klimov-paul 
This problem is not bug, but, I think this is the yii2-httpclient component of the Response getParser method defects. The Response detectFormatByContent method parses the data content only in JSON, URLEncode, and XML formats, but what if the returned content format is text/html? For example, the type of OpenID interface that QQ gets back is text/html and the format of the data returned is:

```
Http-Code: 200
TWS Server:
Date: Tue, 03 Jan 2017 15:08:16 GMT
Text/html Content-Type:
Content-Length: 83
Close Connection:
No-cache Cache-Control:

Callback ({"client_id": "101318901", "openid" "F2FDA4A1E262F956E37F47B767958A65"});
```

I think that the Response detectFormatByContent method and the JsonParser parse method need to make the following changes

Response detectFormatByContent method:

```
/**
     * Detects response format from raw content.
     * @param string $content raw response content.
     * @return null|string format name, 'null' - if detection failed.
     */
    protected function detectFormatByContent($content)
    {
        if (preg_match('/^\\{.*\\}$/is', $content)) {
            return Client::FORMAT_JSON;
        }
        if (preg_match('/^([^=&])+=[^=&]+(&[^=&]+=[^=&]+)*$/', $content)) {
            return Client::FORMAT_URLENCODED;
        }
        if (preg_match('/^<.*>$/s', $content)) {
            return Client::FORMAT_XML;
        }
        // new
        if (strpos($content, "callback(") === 0) {
            $count = 0;
            $jsonData = preg_replace('/^callback\(\s*(\\{.*\\})\s*\);$/is', '\1', $content, 1, $count);
            if ($count === 1) {
                return Client::FORMAT_JSON;
                // $rawResponse = $jsonData;
                // $contentType = self::CONTENT_TYPE_JSON;
            }
        }
        return null;
    }
```

JsonParser parse method:

```
/**
 * @inheritdoc
 */
public function parse(Response $response)
{
	$content = $response->getContent();
	$count = 0;
    $jsonData = preg_replace('/^callback\(\s*(\\{.*\\})\s*\);$/is', '\1', $content, 1, $count);
    if ($count === 1) {
        $content = $jsonData;
    }
    return Json::decode($content);
}
```

In other words, you need to increase support for the following data types
```
Callback ({"client_id": "101318901", "openid" "F2FDA4A1E262F956E37F47B767958A65"});
```

**Comment by manks0127** [comment]
In version 2.0, I can do this in the widget processResponse method, but in the 2.1 version, I have the above problem,

In version 2.0
```
/**
     * @inheritdoc
     */
    protected function processResponse($rawResponse, $contentType = self::CONTENT_TYPE_AUTO)
    {
        if ($contentType === self::CONTENT_TYPE_AUTO && strpos($rawResponse, "callback(") === 0) {
            $count = 0;
            $jsonData = preg_replace('/^callback\(\s*(\\{.*\\})\s*\);$/is', '\1', $rawResponse, 1, $count);
            if ($count === 1) {
                $rawResponse = $jsonData;
                $contentType = self::CONTENT_TYPE_JSON;
            }
        }
        return parent::processResponse($rawResponse, $contentType);
    }
```
If this type of data is not supported by yii2-httpclient
Callback ({"client_id": "101318901", "openid" "F2FDA4A1E262F956E37F47B767958A65"});
I had to rewrite the source code in my home

**Comment by cebe** [comment]
how is a client supposed to extract data from text/html?

**Comment by cebe** [comment]
If you have a special data type like this one, you should not rely on auto detection but specify it explicitly.

**Comment by manks0127** [comment]
@cebe ,in the 2.1 version,Judgment is done in yii2-httpclient's Response, do not support a special data type, I have no way

**Comment by cebe** [comment]
can you show the code your are using?

**Comment by manks0127** [comment]
@cebe ,I simulated a similar request，like this:
```
public function actionIndex()
{

    $client = new \yii\httpclient\Client();
    $response = $client->get('http://opens.imread.com/index.php/site/test222')->send();
    var_dump($response->getData());
    die;
}

public function actionTest222 ()
{
    $data = 'Callback ({"client_id": "101318901", "openid" "F2FDA4A1E262F956E37F47B767958A65"});';
    exit($data);
}
```
The result error is the same.

QQ platform to get the user OpenID_OAuth2.0 returns the result is this
The following is a document corresponding to the QQ,you can look at the interface
Document address:http://wiki.open.qq.com/wiki/website/%E8%8E%B7%E5%8F%96%E7%94%A8%E6%88%B7OpenID_OAuth2.0
Interface:https://graph.qq.com/oauth2.0/me
the return package is as follows:
```
callback( {"client_id":"YOUR_APPID","openid":"YOUR_OPENID"} ); 
```

**Comment by klimov-paul** [comment]
https://github.com/yiisoft/yii2-httpclient/issues/67#issuecomment-243513842

**Comment by klimov-paul** [comment]
You should not use `sendRequest()` then.

**Comment by manks0127** [comment]
You mean this is yii2-authclient's Bug?

**Comment by klimov-paul** [comment]
It is not a bug. It is your ignorance.

**Comment by manks0127** [comment]
This is my code
```
<?php
namespace frontend\components;

use yii\authclient\OAuth2;
use yii\web\HttpException;
use Yii;

class AuthClientQq extends OAuth2
{
    /**
     * @inheritdoc
     */
    public $authUrl = 'https://graph.qq.com/oauth2.0/authorize';
    /**
     * @inheritdoc
     */
    public $tokenUrl = 'https://graph.qq.com/oauth2.0/token';
    /**
     * @inheritdoc
     */
    public $apiBaseUrl = 'https://graph.qq.com';
    
    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
        if ($this->scope === null) {
            $this->scope = implode(' ', [
                'get_user_info',
            ]);
        }
    }
    protected function initUserAttributes()
    {
        $user = $this->api('oauth2.0/me', 'GET');
        if ( isset($user['error']) ) {
            throw new HttpException(400, $user['error']. ':'. $user['error_description']);
        }
        $userAttributes = $this->api(
            "user/get_user_info",
            'GET',
            [
                'oauth_consumer_key' => $user['client_id'],
                'openid' => $user['openid'],
            ]
        );
        yii::warning($userAttributes, 'userAttributes');
        $userAttributes['id'] = $user['openid'];
        $userAttributes['login'] = $userAttributes['nickname'];
        return $userAttributes;
    }
    
    /**
     * @inheritdoc
     */
    protected function defaultName()
    {
        return 'qq';
    }
    /**
     * @inheritdoc
     */
    protected function defaultTitle()
    {
        return 'QQ';
    }
}
```

I hope you can point out my shortcomings

**Comment by klimov-paul** [comment]
I do not have time for your education or programming for you. Ask around forum for the help.

**Comment by manks0127** [comment]
ok,Thanks.
